### PR TITLE
Einav chat extra

### PR DESCRIPTION
**Pull request by nushicoco**
* transition when msg arrives
* system msg when tech arrives the chat room
* indication that a user is typing (which stops when any msg arrives)

### CODE REVIEW

File: client/src/components/chat/chatTab.js [hunk 1]
```diff
@@ -22,16 +24,17 @@ export default class chatTab extends Component {
...
         this.socket = props.socket;
 
-        let self = this;
-
-        getChat(this.props.chatId).then(function (chat) {
-            self.setState({chatClientName:chat.clientName})
-            self.setState({messages:chat.chat_messages});
+        getChat(this.props.chatId).then((chat) => {
+            this.setState({chatClientName: chat.clientName})
+            this.setState({messages: chat.chat_messages})
```
Comment (dancar) [inline client/src/components/chat/chatTab.js]:
```
this.state({
     chatclientName: (...)
     messages: (...)
})
```
?

File: client/src/components/chat/chatTab.js [hunk 2]
```diff
@@ -51,7 +54,37 @@ export default class chatTab extends Component {
...
+
+            if (!this.ignoredOwnConnectMessage){
+                this.ignoredOwnConnectMessage = true
+                return
+            }
+
+            let newUserMsg = {
+                clientName: "system",
+                text: `${data.clientName} ${Strings.chat.hasJoined}`,
+                createdAd: Moment(),
```
Comment (dancar) [inline client/src/components/chat/chatTab.js]: is this suppose to be `createdAt` ?

File: client/src/components/chat/chatTab.js [hunk 3]
```diff
@@ -51,7 +54,37 @@ export default class chatTab extends Component {
...
+                isMe: false
+            }
+
+            this.setState({messages: this.state.messages.concat([ newUserMsg ])})
+        })
+
+        this.socket.on('server:clientIsTyping', data => {
+            if (data.clientId !== this.state.clientId){
+                this.setState({showIsTyping:`${data.clientName} ${Strings.chat.isTyping}`});
```
Comment (dancar) [inline client/src/components/chat/chatTab.js]: the state should best only hold data, not formatting; can you move the formatting/view logic to where rendering occurs ?

File: client/src/components/chat/chatTab.js [hunk 4]
```diff
@@ -51,7 +54,37 @@ export default class chatTab extends Component {
...
+                setTimeout( () => {
```
Comment (dancar) [inline client/src/components/chat/chatTab.js]: say another 'isTyping' event arrives 4.9 seconds from now. won't this timeout inhibit it?

File: client/src/components/chat/chatTab.js [hunk 5]
```diff
@@ -72,21 +105,27 @@ export default class chatTab extends Component {
...
+            <h2>
+                {
+                    this.props.isSuperuser
+                    ? `${Strings.chat.chatWith} ${this.state.chatClientName}`
+                    : Strings.chat.chatWithTech
+                }
+            </h2>
 
             <div className="chat-area" >
+                {this.state.showIsTyping !== '' && (<div><p className="system-message">{this.state.showIsTyping}</p></div>)}
```
Comment (dancar) [inline client/src/components/chat/chatTab.js]: say can you remind me what do we get from adding a <p>, that can't go into the div+class?

File: client/src/components/chat/chatTab.js [hunk 6]
```diff
@@ -51,7 +54,37 @@ export default class chatTab extends Component {
...
+
+            this.setState( {
+                showIsTyping: '',
+                messages : self.state.messages.concat([data])
+            })
+        })
+
+        this.socket.on(`server:userConnected`, data => {
+
+            if (!this.ignoredOwnConnectMessage){
```
Comment (dancar) [inline client/src/components/chat/chatTab.js]: Alternatively, is it possible to emit the isTyping event with an id, and the server won't send this event back to where it came from?

File: client/src/components/chat/chatTab.js (hunk 1 above)
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: oh yeah :) 
11 pm work again..

File: client/src/components/chat/chatTab.js (hunk 6 above)
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: yeah totally

File: client/src/components/chat/chatTab.js (hunk 2 above)
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: lols 
yes

File: client/src/components/chat/chatTab.js (hunk 3 above)
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: sures

File: client/src/components/chat/chatTab.js (hunk 4 above)
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: yeah
not sure it's that big of an issue though.
how would u go about it? have a is typing for every chat member?

File: client/src/components/chat/chatTab.js (hunk 5 above)
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: it's just a habit I guess, using divs only as containers

File: client/src/components/chat/chatTab.js (hunk 1 above)
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: will be fixed in the standard branch

File: client/src/components/chat/chatTab.js (hunk 6 above)
Comment (nushicoco) [inline client/src/components/chat/chatTab.js]: fixed in standard branch

File: client/src/components/chat/chatTab.js (hunk 4 above)
Comment (dancar) [inline client/src/components/chat/chatTab.js]: hmm maybe some like save the timeout reference (setTimeout's return value) and do a clearTimeout() on it when you get a new event?

Comment (dancar) [inline client/src/components/chat/chatTab.js]: Also, consider extracting that 5000 to some prominent, named constant

**Pull request merged**